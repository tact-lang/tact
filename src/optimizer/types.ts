import { AstExpression } from "../grammar/ast";

export abstract class ExpressionTransformer {
    public abstract applyRules(ast: AstExpression): AstExpression;
}

export abstract class Rule {
    public abstract applyRule(
        ast: AstExpression,
        optimizer: ExpressionTransformer,
    ): AstExpression;
}
