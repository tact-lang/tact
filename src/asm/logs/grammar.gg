vmLine =
      VmLoc
    / VmStack
    / VmExecute
    / VmLimitChanged
    / VmGasRemaining
    / VmException
    / VmExceptionHandler
    / VmFinalC5
    / VmUnknown;

VmStack            = "stack: " stack:$[^\n]+;
VmLoc              = "code cell hash:" hash:hex "offset:" offset:number;
VmExecute          = "execute " instr:$[^\n]+;
VmLimitChanged     = "changing gas limit to " limit:number;
VmGasRemaining     = "gas remaining: " gas:number;
VmException        = "handling exception code " errno:number ": " message:$[^\n]*;
VmExceptionHandler = "default exception handler, terminating vm with exit code" errno:number;
VmFinalC5          = "final c5:" value:Cell;
VmUnknown          = !"stack" text:$[^\n]+ "\n";

VmParsedStack = "[" values:VmStackValue* "]";

VmStackValue = value:(
      Null
    / NaN
    / Integer
    / Tuple
    / TupleParen
    / Cell
    / Continuation
    / Builder
    / CellSlice
    / Unknown
);

Null         = "()" / "(null)";
NaN          = "NaN";
Integer      = value:number;
Tuple        = "[" elements:VmStackValue* "]";
TupleParen   = "(" elements:VmStackValue* ")";
Cell         = "C{" value:hex "}";
Continuation = "Cont{" value:$#([A-Za-z_0-9]*) "}";
Builder      = "BC{" value:hex "}";
Unknown      = "???";

CellSlice     = "CS{" body:(CellSliceBody / CellSliceShortBody) "}";
CellSliceBody = "Cell{" value:hex "}" bits:CellSliceBits ";" refs:CellSliceRefs;
CellSliceBits = "bits:" start:number ".." end:number;
CellSliceRefs = "refs:" start:number ".." end:number;

CellSliceShortBody = value:hex;

number   = op:"-"? value:#$(digit+);
digit    = [0-9];

hexDigit "hexadecimal digit" = [0-9a-fA-F];
hex      = $#(value:hexDigit*);

space "space" = [ \t\r\n];
