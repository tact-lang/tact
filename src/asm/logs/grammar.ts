/* Generated. Do not edit. */
/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-redundant-type-constituents */
/* eslint-disable @typescript-eslint/no-duplicate-type-constituents */
/* eslint-disable @typescript-eslint/no-unused-vars */
import * as $ from "@tonstudio/parser-runtime"
export namespace $ast {
    export type VmLoc = $.Located<{
        readonly $: "VmLoc"
        readonly hash: hex
        readonly offset: $number
    }>
    export type VmStack = $.Located<{
        readonly $: "VmStack"
        readonly stack: string
    }>
    export type VmExecute = $.Located<{
        readonly $: "VmExecute"
        readonly instr: string
    }>
    export type VmLimitChanged = $.Located<{
        readonly $: "VmLimitChanged"
        readonly limit: $number
    }>
    export type VmGasRemaining = $.Located<{
        readonly $: "VmGasRemaining"
        readonly gas: $number
    }>
    export type VmException = $.Located<{
        readonly $: "VmException"
        readonly errno: $number
        readonly message: string
    }>
    export type VmExceptionHandler = $.Located<{
        readonly $: "VmExceptionHandler"
        readonly errno: $number
    }>
    export type VmFinalC5 = $.Located<{
        readonly $: "VmFinalC5"
        readonly value: Cell
    }>
    export type VmUnknown = $.Located<{
        readonly $: "VmUnknown"
        readonly text: string
    }>
    export type vmLine =
        | VmLoc
        | VmStack
        | VmExecute
        | VmLimitChanged
        | VmGasRemaining
        | VmException
        | VmExceptionHandler
        | VmFinalC5
        | VmUnknown
    export type VmParsedStack = $.Located<{
        readonly $: "VmParsedStack"
        readonly values: readonly VmStackValue[]
    }>
    export type VmStackValue = $.Located<{
        readonly $: "VmStackValue"
        readonly value:
            | Null
            | NaN
            | Integer
            | Tuple
            | TupleParen
            | Cell
            | Continuation
            | Builder
            | CellSlice
            | Unknown
    }>
    export type Null = $.Located<{
        readonly $: "Null"
    }>
    export type NaN = $.Located<{
        readonly $: "NaN"
    }>
    export type Integer = $.Located<{
        readonly $: "Integer"
        readonly value: $number
    }>
    export type Tuple = $.Located<{
        readonly $: "Tuple"
        readonly elements: readonly VmStackValue[]
    }>
    export type TupleParen = $.Located<{
        readonly $: "TupleParen"
        readonly elements: readonly VmStackValue[]
    }>
    export type Cell = $.Located<{
        readonly $: "Cell"
        readonly value: hex
    }>
    export type Continuation = $.Located<{
        readonly $: "Continuation"
        readonly value: string
    }>
    export type Builder = $.Located<{
        readonly $: "Builder"
        readonly value: hex
    }>
    export type Unknown = $.Located<{
        readonly $: "Unknown"
    }>
    export type CellSlice = $.Located<{
        readonly $: "CellSlice"
        readonly body: CellSliceBody | CellSliceShortBody
    }>
    export type CellSliceBody = $.Located<{
        readonly $: "CellSliceBody"
        readonly value: hex
        readonly bits: CellSliceBits
        readonly refs: CellSliceRefs
    }>
    export type CellSliceBits = $.Located<{
        readonly $: "CellSliceBits"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceRefs = $.Located<{
        readonly $: "CellSliceRefs"
        readonly start: $number
        readonly end: $number
    }>
    export type CellSliceShortBody = $.Located<{
        readonly $: "CellSliceShortBody"
        readonly value: hex
    }>
    export type $number = {
        readonly op: "-" | undefined
        readonly value: string
    }
    export type digit = string
    export type hexDigit = string | string | string
    export type hex = string
    export type space = " " | "\t" | "\r" | "\n"
}
export const VmLoc: $.Parser<$ast.VmLoc> = $.loc(
    $.field(
        $.pure("VmLoc"),
        "$",
        $.right(
            $.str("code cell hash:"),
            $.field(
                $.lazy(() => hex),
                "hash",
                $.right(
                    $.str("offset:"),
                    $.field(
                        $.lazy(() => $number),
                        "offset",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmStack: $.Parser<$ast.VmStack> = $.loc(
    $.field(
        $.pure("VmStack"),
        "$",
        $.right(
            $.str("stack: "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "stack",
                $.eps,
            ),
        ),
    ),
)
export const VmExecute: $.Parser<$ast.VmExecute> = $.loc(
    $.field(
        $.pure("VmExecute"),
        "$",
        $.right(
            $.str("execute "),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "instr",
                $.eps,
            ),
        ),
    ),
)
export const VmLimitChanged: $.Parser<$ast.VmLimitChanged> = $.loc(
    $.field(
        $.pure("VmLimitChanged"),
        "$",
        $.right(
            $.str("changing gas limit to "),
            $.field(
                $.lazy(() => $number),
                "limit",
                $.eps,
            ),
        ),
    ),
)
export const VmGasRemaining: $.Parser<$ast.VmGasRemaining> = $.loc(
    $.field(
        $.pure("VmGasRemaining"),
        "$",
        $.right(
            $.str("gas remaining: "),
            $.field(
                $.lazy(() => $number),
                "gas",
                $.eps,
            ),
        ),
    ),
)
export const VmException: $.Parser<$ast.VmException> = $.loc(
    $.field(
        $.pure("VmException"),
        "$",
        $.right(
            $.str("handling exception code "),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.right(
                    $.str(": "),
                    $.field(
                        $.stry($.star($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                        "message",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const VmExceptionHandler: $.Parser<$ast.VmExceptionHandler> = $.loc(
    $.field(
        $.pure("VmExceptionHandler"),
        "$",
        $.right(
            $.str("default exception handler, terminating vm with exit code"),
            $.field(
                $.lazy(() => $number),
                "errno",
                $.eps,
            ),
        ),
    ),
)
export const VmFinalC5: $.Parser<$ast.VmFinalC5> = $.loc(
    $.field(
        $.pure("VmFinalC5"),
        "$",
        $.right(
            $.str("final c5:"),
            $.field(
                $.lazy(() => Cell),
                "value",
                $.eps,
            ),
        ),
    ),
)
export const VmUnknown: $.Parser<$ast.VmUnknown> = $.loc(
    $.field(
        $.pure("VmUnknown"),
        "$",
        $.right(
            $.lookNeg($.str("stack")),
            $.field(
                $.stry($.plus($.regex<"\n">("^\\n", $.negateExps([$.ExpString("\n")])))),
                "text",
                $.right($.str("\n"), $.eps),
            ),
        ),
    ),
)
export const vmLine: $.Parser<$ast.vmLine> = $.alt(
    VmLoc,
    $.alt(
        VmStack,
        $.alt(
            VmExecute,
            $.alt(
                VmLimitChanged,
                $.alt(
                    VmGasRemaining,
                    $.alt(VmException, $.alt(VmExceptionHandler, $.alt(VmFinalC5, VmUnknown))),
                ),
            ),
        ),
    ),
)
export const VmParsedStack: $.Parser<$ast.VmParsedStack> = $.loc(
    $.field(
        $.pure("VmParsedStack"),
        "$",
        $.right(
            $.str("["),
            $.field($.star($.lazy(() => VmStackValue)), "values", $.right($.str("]"), $.eps)),
        ),
    ),
)
export const VmStackValue: $.Parser<$ast.VmStackValue> = $.loc(
    $.field(
        $.pure("VmStackValue"),
        "$",
        $.field(
            $.alt(
                $.lazy(() => Null),
                $.alt(
                    $.lazy(() => NaN),
                    $.alt(
                        $.lazy(() => Integer),
                        $.alt(
                            $.lazy(() => Tuple),
                            $.alt(
                                $.lazy(() => TupleParen),
                                $.alt(
                                    $.lazy(() => Cell),
                                    $.alt(
                                        $.lazy(() => Continuation),
                                        $.alt(
                                            $.lazy(() => Builder),
                                            $.alt(
                                                $.lazy(() => CellSlice),
                                                $.lazy(() => Unknown),
                                            ),
                                        ),
                                    ),
                                ),
                            ),
                        ),
                    ),
                ),
            ),
            "value",
            $.eps,
        ),
    ),
)
export const Null: $.Parser<$ast.Null> = $.loc(
    $.field($.pure("Null"), "$", $.right($.alt($.str("()"), $.str("(null)")), $.eps)),
)
export const NaN: $.Parser<$ast.NaN> = $.loc(
    $.field($.pure("NaN"), "$", $.right($.str("NaN"), $.eps)),
)
export const Integer: $.Parser<$ast.Integer> = $.loc(
    $.field(
        $.pure("Integer"),
        "$",
        $.field(
            $.lazy(() => $number),
            "value",
            $.eps,
        ),
    ),
)
export const Tuple: $.Parser<$ast.Tuple> = $.loc(
    $.field(
        $.pure("Tuple"),
        "$",
        $.right($.str("["), $.field($.star(VmStackValue), "elements", $.right($.str("]"), $.eps))),
    ),
)
export const TupleParen: $.Parser<$ast.TupleParen> = $.loc(
    $.field(
        $.pure("TupleParen"),
        "$",
        $.right($.str("("), $.field($.star(VmStackValue), "elements", $.right($.str(")"), $.eps))),
    ),
)
export const Cell: $.Parser<$ast.Cell> = $.loc(
    $.field(
        $.pure("Cell"),
        "$",
        $.right(
            $.str("C{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Continuation: $.Parser<$ast.Continuation> = $.loc(
    $.field(
        $.pure("Continuation"),
        "$",
        $.right(
            $.str("Cont{"),
            $.field(
                $.stry(
                    $.lex(
                        $.star(
                            $.regex<string | string | "_" | string>("A-Za-z_0-9", [
                                $.ExpRange("A", "Z"),
                                $.ExpRange("a", "z"),
                                $.ExpString("_"),
                                $.ExpRange("0", "9"),
                            ]),
                        ),
                    ),
                ),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Builder: $.Parser<$ast.Builder> = $.loc(
    $.field(
        $.pure("Builder"),
        "$",
        $.right(
            $.str("BC{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const Unknown: $.Parser<$ast.Unknown> = $.loc(
    $.field($.pure("Unknown"), "$", $.right($.str("???"), $.eps)),
)
export const CellSlice: $.Parser<$ast.CellSlice> = $.loc(
    $.field(
        $.pure("CellSlice"),
        "$",
        $.right(
            $.str("CS{"),
            $.field(
                $.alt(
                    $.lazy(() => CellSliceBody),
                    $.lazy(() => CellSliceShortBody),
                ),
                "body",
                $.right($.str("}"), $.eps),
            ),
        ),
    ),
)
export const CellSliceBody: $.Parser<$ast.CellSliceBody> = $.loc(
    $.field(
        $.pure("CellSliceBody"),
        "$",
        $.right(
            $.str("Cell{"),
            $.field(
                $.lazy(() => hex),
                "value",
                $.right(
                    $.str("}"),
                    $.field(
                        $.lazy(() => CellSliceBits),
                        "bits",
                        $.right(
                            $.str(";"),
                            $.field(
                                $.lazy(() => CellSliceRefs),
                                "refs",
                                $.eps,
                            ),
                        ),
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceBits: $.Parser<$ast.CellSliceBits> = $.loc(
    $.field(
        $.pure("CellSliceBits"),
        "$",
        $.right(
            $.str("bits:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceRefs: $.Parser<$ast.CellSliceRefs> = $.loc(
    $.field(
        $.pure("CellSliceRefs"),
        "$",
        $.right(
            $.str("refs:"),
            $.field(
                $.lazy(() => $number),
                "start",
                $.right(
                    $.str(".."),
                    $.field(
                        $.lazy(() => $number),
                        "end",
                        $.eps,
                    ),
                ),
            ),
        ),
    ),
)
export const CellSliceShortBody: $.Parser<$ast.CellSliceShortBody> = $.loc(
    $.field(
        $.pure("CellSliceShortBody"),
        "$",
        $.field(
            $.lazy(() => hex),
            "value",
            $.eps,
        ),
    ),
)
export const $number: $.Parser<$ast.$number> = $.field(
    $.opt($.str("-")),
    "op",
    $.field($.lex($.stry($.plus($.lazy(() => digit)))), "value", $.eps),
)
export const digit: $.Parser<$ast.digit> = $.regex<string>("0-9", [$.ExpRange("0", "9")])
export const hexDigit: $.Parser<$ast.hexDigit> = $.named(
    "hexadecimal digit",
    $.regex<string | string | string>("0-9a-fA-F", [
        $.ExpRange("0", "9"),
        $.ExpRange("a", "f"),
        $.ExpRange("A", "F"),
    ]),
)
export const hex: $.Parser<$ast.hex> = $.stry($.lex($.field($.star(hexDigit), "value", $.eps)))
export const space: $.Parser<$ast.space> = $.named(
    "space",
    $.regex<" " | "\t" | "\r" | "\n">(" \\t\\r\\n", [
        $.ExpString(" "),
        $.ExpString("\t"),
        $.ExpString("\r"),
        $.ExpString("\n"),
    ]),
)
