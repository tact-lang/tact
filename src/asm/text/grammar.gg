File = instructions:instructions;

instructions = (Instruction / ExplicitRef / EmbedSlice / Exotic)*;

Instruction = name:Id args:Argument*;

Argument = expression:(
      IntegerLiteral
    / DataLiteral
    / Code
    / Dictionary
    / StackElement
    / ControlRegister
);

StackElement = value:$#("s" "-"? IntegerLiteralDec);
ControlRegister = value:$#("c" IntegerLiteralDec);

Code = "{" instructions:instructions "}";

Dictionary = "[" entries:DictionaryEntry* "]";

DictionaryEntry = id:IntegerLiteral "=>" code:Code;

ExplicitRef = keyword<"ref"> code:Code;
EmbedSlice  = keyword<"embed"> data:DataLiteral;
Exotic      = keyword<"exotic"> lib:(ExoticLibrary / DefaultExotic);

ExoticLibrary = keyword<"library"> data:DataLiteral;
DefaultExotic = data:DataLiteral;

idPart "identifier character" = [a-zA-Z0-9_];
Id "identifier" = name:#$(!reservedWord [a-zA-Z_] idPart*);

DataLiteral = value:(HexLiteral / BinLiteral / BocLiteral / StringLiteral);

HexLiteral = "x{" content:$(hexDigit* "_"?) "}";
BinLiteral = "b{" content:$(binDigit*) "}";
BocLiteral = "boc{" content:$(hexDigit*) "}";

// Literals
IntegerLiteral = op:"-"? value:(IntegerLiteralHex / IntegerLiteralBin / IntegerLiteralOct / IntegerLiteralDec);

IntegerLiteralDec = digits:#underscored<digit>;
IntegerLiteralHex = digits:#("0" [xX] @underscored<hexDigit>);
IntegerLiteralBin = digits:#("0" [bB] @underscored<binDigit>);
IntegerLiteralOct = digits:#("0" [oO] @underscored<[0-7]>);

underscored<T> = $(T ("_"? T)*);

hexDigit "hexadecimal digit" = [0-9a-fA-F];
binDigit "binary digit" = [01];
digit "digit" = [0-9];

StringLiteral = value:#("\"" @$([^"\\])* "\"");

keyword<T> = #(@T !idPart);

reservedWord "reserved word" = keyword<(
    "ref" / "embed" / "exotic" / "library"
)>;

space "space" = [ \t\r\n] / comment;
comment = singleLineComment;
singleLineComment = "//" @$[^\r\n]*;
