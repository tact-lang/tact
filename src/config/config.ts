export type SafetyOptions = {
    /**
     * If set to `true`, enables run-time null checks for the `!!` operator. Default is `true`.
     */
    readonly nullChecks?: boolean;
};

export type ExperimentalOptions = {
    /**
     * If set to true, enables inlining of all functions in contracts.
     * This can reduce gas usage at the cost of bigger contracts.
     */
    readonly inline?: boolean;
};

/**
 * Per-project configuration options
 *
 * Read more: https://docs.tact-lang.org/book/config#projects
 */
export type Options = {
    /**
     * If set to true, enables debug output of a contract and allows usage of `dump()` function,
     * which is useful for debugging purposes.
     *
     * Read more: https://docs.tact-lang.org/book/debug
     */
    readonly debug?: boolean;
    /**
     * If set to true, enables support of external message receivers.
     *
     * Read more: https://docs.tact-lang.org/book/external
     */
    readonly external?: boolean;
    /**
     * If set to true, enables generation of a getter with IPFS links describing the contract's ABI.
     *
     * Read more: https://docs.tact-lang.org/ref/evolution/otp-003
     */
    readonly ipfsAbiGetter?: boolean;
    /**
     * If set to true, enables generation of a getter with a list of interfaces provided by the contract.
     *
     * Read more: https://docs.tact-lang.org/book/contracts#interfaces
     */
    readonly interfacesGetter?: boolean;
    /**
     * If set to "new", uses new parser. If set to "old", uses legacy parser. Default is "old".
     */
    readonly parser?: "new" | "old";
    /**
     * Experimental options that might be removed in the future. Use with caution!
     */
    readonly experimental?: ExperimentalOptions;
    /**
     * Safety options for the contract.
     */
    readonly safety?: SafetyOptions;
    /**
     * If set to true, enables generation of `lazy_deployment_completed()` getter.
     */
    readonly enableLazyDeploymentCompletedGetter?: boolean;
};

export type Mode = "fullWithDecompilation" | "full" | "funcOnly" | "checkOnly";

/**
 * Per-project configuration options
 *
 * Read more: https://docs.tact-lang.org/book/config#projects
 */
export type Project = {
    /**
     * Name of the project. All generated files are prefixed with it.
     *
     * Read more: https://docs.tact-lang.org/book/config#projects-name
     */
    name: string;
    /**
     * Path to the project's Tact file. You can only specify one Tact file per project.
     *
     * Read more: https://docs.tact-lang.org/book/config#projects-path
     */
    path: string;
    /**
     * Path to the directory where all generated files will be placed.
     *
     * Read more: https://docs.tact-lang.org/book/config#projects-output
     */
    output: string;
    /**
     * Compilation options for the project.
     *
     * Read more: https://docs.tact-lang.org/book/config#projects-options
     */
    options?: Options;
    /**
     * Compilation mode of the project.
     *
     * Read more: https://docs.tact-lang.org/book/config#projects-mode
     */
    mode?: Mode;

    /**
     * Set verbosity level (higher = more details), default: 1
     */
    verbose?: number;
};

/**
 * Compiler configuration schema
 *
 * Read more: https://docs.tact-lang.org/book/config
 */
export type Config = {
    /**
     * A property for specifying a path or URL to the JSON schema of tact.config.json
     *
     * Read more: https://docs.tact-lang.org/book/config#schema
     */
    $schema?: string;
    /**
     * List of Tact projects with respective compilation options. Each .tact file represents its own Tact project.
     *
     * Read more: https://docs.tact-lang.org/book/config#projects
     */
    projects: readonly Project[];
};
