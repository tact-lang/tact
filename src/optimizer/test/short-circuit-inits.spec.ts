import { getAstFactory } from "../../ast/ast-helpers";
import { featureEnable } from "../../config/features";
import { CompilerContext } from "../../context/context";
import { openContext } from "../../context/store";
import { getParser } from "../../grammar";
import { defaultParser } from "../../grammar/grammar";
import { resolveDescriptors } from "../../types/resolveDescriptors";
import { getAllExpressionTypes } from "../../types/resolveExpression";
import { resolveSignatures } from "../../types/resolveSignatures";
import { resolveStatements } from "../../types/resolveStatements";
import { loadCases } from "../../utils/loadCases";

describe("constant-propagation", () => {
    for (const r of loadCases(__dirname + "/short-circuit/success/")) {
        it(
            r.name +
                " should pass boolean short-circuiting during initialization",
            () => {
                const Ast = getAstFactory();
                let ctx = openContext(
                    new CompilerContext(),
                    [{ code: r.code, path: "<unknown>", origin: "user" }],
                    [],
                    getParser(Ast, defaultParser),
                );
                ctx = featureEnable(ctx, "external");
                ctx = resolveDescriptors(ctx, Ast);
                ctx = resolveStatements(ctx, Ast);
                ctx = resolveSignatures(ctx, Ast);
                expect(getAllExpressionTypes(ctx)).toMatchSnapshot();
            },
        );
    }
    for (const r of loadCases(__dirname + "/short-circuit/failed/")) {
        it(
            r.name +
                " should fail boolean short-circuiting during initialization",
            () => {
                const Ast = getAstFactory();
                let ctx = openContext(
                    new CompilerContext(),
                    [{ code: r.code, path: "<unknown>", origin: "user" }],
                    [],
                    getParser(Ast, defaultParser),
                );
                ctx = featureEnable(ctx, "external");
                expect(() => {
                    ctx = resolveDescriptors(ctx, Ast);
                    ctx = resolveStatements(ctx, Ast);
                    ctx = resolveSignatures(ctx, Ast);
                }).toThrowErrorMatchingSnapshot();
            },
        );
    }
});
