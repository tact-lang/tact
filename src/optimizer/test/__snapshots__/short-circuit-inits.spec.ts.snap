// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constant-propagation constant-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:4:35: Cannot evaluate expression to a constant: divisor expression must be non-zero
  3 | 
> 4 | const A: Bool = true && exception(0);
                                        ^
  5 | 
"
`;

exports[`constant-propagation constant-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:4:36: Cannot evaluate expression to a constant: divisor expression must be non-zero
  3 | 
> 4 | const A: Bool = false || exception(0);
                                         ^
  5 | 
"
`;

exports[`constant-propagation contract-constant-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:39: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     const A: Bool = true && exception(0);
                                             ^
  13 | }
"
`;

exports[`constant-propagation contract-constant-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:40: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     const A: Bool = false || exception(0);
                                              ^
  13 | }
"
`;

exports[`constant-propagation contract-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:33: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     A: Bool = true && exception(0);
                                       ^
  13 | }
"
`;

exports[`constant-propagation contract-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:12:34: Cannot evaluate expression to a constant: divisor expression must be non-zero
  11 | contract CTest {
> 12 |     A: Bool = false || exception(0);
                                        ^
  13 | }
"
`;

exports[`constant-propagation short-circuit-initializations should pass boolean short-circuiting during initialization 1`] = `
[
  [
    "true",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "true || loop()",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "true || exception(0)",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "false && loop()",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "false && exception(0)",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "true || loop()",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "true || exception(0)",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "false && loop()",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "false && exception(0)",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "true || loop()",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "true || exception(0)",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "false && loop()",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "false && exception(0)",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "true || loop()",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "true || exception(0)",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "loop()",
    "Bool",
  ],
  [
    "false && loop()",
    "Bool",
  ],
  [
    "false",
    "Bool",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exception(0)",
    "Bool",
  ],
  [
    "false && exception(0)",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "1",
    "Int",
  ],
  [
    "v",
    "Int",
  ],
  [
    "1 / v",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "1 / v == 0",
    "Bool",
  ],
]
`;

exports[`constant-propagation struct-init-and should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:5:33: Cannot evaluate expression to a constant: divisor expression must be non-zero
  4 | struct STest {
> 5 |     A: Bool = true && exception(0);
                                      ^
  6 | }
"
`;

exports[`constant-propagation struct-init-or should fail boolean short-circuiting during initialization 1`] = `
"<unknown>:5:34: Cannot evaluate expression to a constant: divisor expression must be non-zero
  4 | struct STest {
> 5 |     A: Bool = false || exception(0);
                                       ^
  6 | }
"
`;
